elasticdl/pkg/common/vector.go

```diff
@@ -0,0 +1,235 @@
+var Int32Dtype = DataType{reflect.TypeOf([]int32{0}), 4, 3}
+
+// Int64Dtype predefined DataType for []int64
+var Int64Dtype = DataType{reflect.TypeOf([]int64{0}), 8, 4}
+
+// Float16Dtype predefined DataType for []float16
+var Float16Dtype = DataType{nil, 0, 5}
+
+// Float32Dtype predefined DataType for []float32
+var Float32Dtype = DataType{reflect.TypeOf([]float32{0}), 4, 6}
+
+// Float64Dtype predefined DataType for []float64
+var Float64Dtype = DataType{reflect.TypeOf([]float64{0}), 8, 7}
+
+// BoolDtype predefined DataType for Bool
+var BoolDtype = DataType{nil, 0, 8}
+
+// TypeToDataType golang reflect.Type -> DataType
+var TypeToDataType = make(map[reflect.Type]DataType)
+
+func init() {
+	TypeToDataType[Int8Dtype.Type] = Int8Dtype
+	TypeToDataType[Int16Dtype.Type] = Int16Dtype
+	TypeToDataType[Int32Dtype.Type] = Int32Dtype
+	TypeToDataType[Int64Dtype.Type] = Int64Dtype
+	TypeToDataType[Float32Dtype.Type] = Float32Dtype
+	TypeToDataType[Float64Dtype.Type] = Float64Dtype
+}
+
+// Vector definition
+type Vector struct {
+	Data   []byte
+	Length int
+	Dtype  DataType
+}
+
+// InitialzeFunc func
+type InitialzeFunc func([]byte) error
+
+// ZeroInit return a zero initializer
+func ZeroInit() InitialzeFunc {
+	return func(buffer []byte) error {
+		// do nothing since golang has already zero init
+		return nil
+	}
+}
+
+func byteSet(buffer []byte, idx int, size int, val interface{}) {
+	switch val.(type) {
+	case int8:
+		buffer[idx] = byte(val.(int8))
+	case int16:
+		binary.LittleEndian.PutUint16(buffer[idx*size:idx*size+size], uint16(val.(int16)))
+	case int32:
+		binary.LittleEndian.PutUint32(buffer[idx*size:idx*size+size], uint32(val.(int32)))
+	case int64:
+		binary.LittleEndian.PutUint64(buffer[idx*size:idx*size+size], uint64(val.(int64)))
+	case float32:
+		binary.LittleEndian.PutUint32(buffer[idx*size:idx*size+size], math.Float32bits(val.(float32)))
+	case float64:
+		binary.LittleEndian.PutUint64(buffer[idx*size:idx*size+size], math.Float64bits(val.(float64)))
+	}
+}
+
+// ConstantInit return a constant iniitializer
+func ConstantInit(val interface{}) InitialzeFunc {
+	return func(buffer []byte) error {
+		size := int(reflect.ValueOf(val).Type().Size())
+		length := len(buffer) / size
+		for idx := 0; idx < length; idx++ {
+			byteSet(buffer, idx, size, val)
+		}
+		return nil
+	}
+}
+
+// CopyInit return a copy initializer
+func CopyInit(slice interface{}) InitialzeFunc {
+	return func(b []byte) error {
+		buffer := bytes.NewBuffer(b)
+		buffer.Reset()
+		return binary.Write(buffer, binary.LittleEndian, slice)
+	}
+}
+
+// NewEmptyVector create an zero-initialized vector
+func NewEmptyVector(length int, dtype DataType) *Vector {
+	bytelen := int(dtype.Size) * length
+	var vec = Vector{
+		Data:   make([]byte, bytelen, bytelen),
+		Dtype:  dtype,
+		Length: length,
+	}
+	return &vec
+}
+
+// NewInitializerVector create an initialized vector
+func NewInitializerVector(length int, dtype DataType, initializer InitialzeFunc) *Vector {
```

Comment:
NewInitializerVector is also a little confusing. Pass a callback initializer function here is also introduce complexity. Maybe we could follow numpy style:

```
NewZeroBuffer --> np.zeros
NewConstantBuffer --> np.full
```

Reply:
the initializer is for future concern, this could support custom init function like Gauss or something else.
In the embedding table definition, there is a field named initializer